File: utilities/DoubleLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class DoubleLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None


    def push_back(self, data):
        new_node = Node(data)

        if self.head is None:
            self.head = new_node
            self.tail = new_node
            return
        
        self.tail.next = new_node
        new_node.prev = self.tail
        self.tail = new_node
        return

    
    def push_front(self, data):
        new_node = Node(data)

        if self.head is None:
            self.head = new_node
            self.tail = new_node
            return
        
        new_node.next = self.head
        self.head.prev = new_node
        self.head = new_node
        return


    def pop_back(self):
        if self.tail is None:
            return None
        
        value = self.tail.data

        if self.head == self.tail:
            self.head = None
            self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None

        return value


    def pop_front(self):
        if self.head is None:
            return None
        
        value = self.head.data
        
        if self.head == self.tail:
            self.head = None
            self.tail = None
        
        else:
            self.head = self.head.next
            self.head.prev = None
        
        return value


    def search(self, data):
        curr = self.head
        while curr is not None:
            if curr.data == data:
                return curr
            curr = curr.next
        
        return None

    
    def delete(self, data):
        node = self.search(data)
        if node is None:
            return False

        if node == self.head:
            self.pop_front()
            return True
        elif node == self.tail:
            self.pop_back()
            return True
        
        node.prev.next = node.next
        node.next.prev = node.prev
        return True


    def display_forward(self):
        curr = self.head
        while curr is not None:
            print(curr.data, end=" <-> ")
            curr = curr.next
        print("None")
        return


    def display_backward(self):
        curr = self.tail
        while curr is not None:
            print(curr.data, end=" <-> ")
            curr = curr.prev
        print("None")
        return


if __name__ == "__main__":

    '''DEBUG'''
    
    dll = DoubleLinkedList()

    dll.push_back(10)
    dll.push_back(20)
    dll.push_back(30)

    dll.push_front(5)

    dll.display_forward()

    dll.display_backward()

    dll.delete(20)

    dll.display_forward()

    print(dll.pop_front())

    print(dll.pop_back())

File: utilities/cpp/DoubleLinkedList.cpp
#include <headers/DoubleLinkedList.hpp>

#include <iostream>
#include <string>

Node::Node(std::string value) {
    data = value;
    next = nullptr;
    prev = nullptr;
}


DoubleLinkedList::DoubleLinkedList() {
    head = nullptr;
    tail = nullptr;
}

void DoubleLinkedList::push_back(std::string data) {
    Node* newNode = new Node(data);

    if(head == nullptr) {
        head = newNode;
        tail = newNode;
        return;
    }

    tail->next = newNode;
    newNode->prev = tail;
    tail = newNode;
    return;
}

void DoubleLinkedList::push_front(std::string data) {
    Node* newNode = new Node(data);

    if(head == nullptr) {
        head = newNode;
        tail = newNode;
        return;
    }

    head->prev = newNode;
    newNode->next = head;
    head = newNode;
    return;
}

void DoubleLinkedList::pop_back() {
    if(tail == nullptr) {
        return;
    }

    Node* temp = tail;

    if(tail == head) {
        head = nullptr;
        tail = nullptr;
        delete temp;
        return;
    }

    tail = tail->prev;
    tail->next = nullptr;
    delete temp;
    return;
}

void DoubleLinkedList::pop_front() {
    if(head == nullptr) {
        return;
    }

    Node* temp = head;

    if(head == tail) {
        head = nullptr;
        tail = nullptr;
        delete temp;
        return;
    }

    head = head->next;
    head->prev = nullptr;
    delete temp;
    return;
}

Node* DoubleLinkedList::search(std::string data) {
    Node* curr = head;
    
    while(curr != nullptr) {
        if(curr->data == data) {
            return curr;
        }
        curr = curr->next;
    }

    return nullptr;
}

void DoubleLinkedList::delete_node(std::string data) {
    Node* reqNode = DoubleLinkedList::search(data);

    if(reqNode == nullptr) {
        return;
    }

    else if(reqNode == head) {
        DoubleLinkedList::pop_front();
        return;
    }

    else if(reqNode == tail) {
        DoubleLinkedList::pop_back();
        return;
    }

    reqNode->prev->next = reqNode->next;
    reqNode->next->prev = reqNode->prev;

    delete reqNode;
    return;
}

void DoubleLinkedList::display_forward() {
    Node* current = head;

    while (current != nullptr) {
        std::cout << current->data << " <-> ";
        current = current->next;
    }

    std::cout << "NULL" << std::endl;
    return;
}

void DoubleLinkedList::display_backward() {
    Node* current = tail;

    while (current != nullptr) {
        std::cout << current->data << " <-> ";
        current = current->prev;
    }

    std::cout << "NULL" << std::endl;
    return;
}

DoubleLinkedList::~DoubleLinkedList() {
    while(head != nullptr) {
        DoubleLinkedList::pop_front();
    }
    
    return;
}